### PR TITLE
fix(progress-circle): render determinate and indeterminate across all supported browsers

### PR DESCRIPTION
**Pull request by josephperrott**
For #295

### CODE REVIEW

File: src/components/progress-circle/progress-circle.scss [hunk 1]
```diff
@@ -3,15 +3,12 @@
 @import "default-theme";
 
 /* Animation Durations */
-$md-progress-circle-duration : 5.25s !default;
-$md-progress-circle-value-change-duration : $md-progress-circle-duration * 0.25 !default;
+$md-progress-circle-duration : 5332ms !default;
```
Comment (jelbourn) [inline src/components/progress-circle/progress-circle.scss]: Can you add any info on where this value comes from?

File: src/components/progress-circle/progress-circle.ts [hunk 2]
```diff
@@ -1,13 +1,19 @@
 import {
   Component,
-  ChangeDetectionStrategy,
   HostBinding,
+  ChangeDetectorRef,
+  ChangeDetectionStrategy,
   Input
 } from 'angular2/core';
-
+import {isPresent} from 'angular2/src/facade/lang';
```
Comment (jelbourn) [inline src/components/progress-circle/progress-circle.ts]: `isPresent` is gone. Just use truthiness or falsiness now.

File: src/components/progress-circle/progress-circle.ts [hunk 3]
```diff
@@ -1,13 +1,19 @@
...
 
 // TODO(josephperrott): Benchpress tests.
 
+let DEGREE_IN_RADIANS = Math.PI / 180;
+let durationIndeterminate: number = 5332 / 8;
+let startIndeterminate = 3;
+let endIndeterminate = 80;
```
Comment (jelbourn) [inline src/components/progress-circle/progress-circle.ts]: Need some comments what what each of these are used for.

File: src/components/progress-circle/progress-circle.ts [hunk 4]
```diff
@@ -25,14 +31,51 @@ import {
...
   @HostBinding('attr.aria-valuenow')
-  @Input('value')
-  _value: number = 0;
+  get value() {
+    return this._value;
+  }
+  set value(v: number) {
+    if (isPresent(v)) {
+      let newValue = MdProgressCircle.clamp(v);
+      this._renderCircle(this.value, newValue, MdProgressCircle.linearEase, 225, 0);
```
Comment (jelbourn) [inline src/components/progress-circle/progress-circle.ts]: Where does the `225` here come from? Move it to a constant?

File: src/components/progress-circle/progress-circle.ts [hunk 5]
```diff
@@ -41,47 +84,172 @@ export class MdProgressCircle {
    * mode is bound to the host as the attribute host.
    */
   @HostBinding('attr.mode')
-  @Input() mode: 'determinate' | 'indeterminate' = 'determinate';
+  @Input()
+  get mode() {
+    return this._mode;
+  }
+  set mode(m: 'determinate' | 'indeterminate') {
```
Comment (jelbourn) [inline src/components/progress-circle/progress-circle.ts]: Rather than recreating the string literal type each time, you can declare it once and reuse it:
```ts
export type ProgressCircleMode = 'determinate' | 'indeterminate';
```

File: src/components/progress-circle/progress-circle.ts [hunk 6]
```diff
@@ -41,47 +84,172 @@ export class MdProgressCircle {
...
+      largeArcFlag = endAngle <= 180 ? 0 : 1;
+    }
+
+    return `M${start}A${pathRadius},${pathRadius} 0 ${largeArcFlag},${arcSweep} ${end}`;
+  }
+
+
   /** Clamps a value to be between 0 and 100. */
-  static clamp(v: number) {
+  private static clamp(v: number) {
```
Comment (jelbourn) [inline src/components/progress-circle/progress-circle.ts]: Rather than being a private static method, you can just have a function that lives in the module. 
```ts
function clamp() {
  ...
}
```

File: src/components/progress-circle/progress-circle.ts [hunk 7]
```diff
@@ -41,47 +84,172 @@ export class MdProgressCircle {
...
+
+
+  /**
+   * Returns the current timestamp either based on the performance global or a date object.
+   */
+  private static _animationNow() {
+    if (window.performance) {
+      return performance.now();
+    }
+    return (new Date).getTime();
```
Comment (jelbourn) [inline src/components/progress-circle/progress-circle.ts]: `Date.now()` is available in all of the browsers that we support.

File: src/components/progress-circle/progress-circle.ts [hunk 8]
```diff
@@ -25,14 +31,51 @@ import {
 })
 export class MdProgressCircle {
   /**
+   * The id of the last requested animation.
+   */
+  private _lastAnimationId: number = 0;
+
+  /**
+   * The id of the indeterminate interval.
```
Comment (jelbourn) [inline src/components/progress-circle/progress-circle.ts]:
```ts
/** The id of the most recent indeterminate setInterval. */
```

File: src/components/progress-circle/progress-circle.ts [hunk 9]
```diff
@@ -41,47 +84,172 @@ export class MdProgressCircle {
...
-   *
-   * The stroke dash offset specifies the distance between dashes in the circle's stroke.
-   * Setting the offset to a percentage of the total circumference of the circle, fills this
-   * percentage of the overall circumference of the circle.
+   * Animates the rendered circle.
    */
-  strokeDashOffset() {
-    // To determine how far the offset should be, we multiple the current percentage by the
-    // total circumference.
+  private _renderCircle(animateFrom: number, animateTo: number, ease: Function,
```
Comment (jelbourn) [inline src/components/progress-circle/progress-circle.ts]: Definitely need a description that explains how this functions works at a high level.

File: src/components/progress-circle/progress-circle.ts [hunk 10]
```diff
@@ -41,47 +84,172 @@ export class MdProgressCircle {
...
-    return 251.3274 * (100 - this._value) / 100;
+        this.currentPath = MdProgressCircle.getSvgArc(
+          ease(elapsedTime, animateFrom, changeInValue, animationDuration),
+          rotation
+        );
+
+        // Prevent overlapping animations by checking if a new animation has been called for and
+        // if the animation has lasted long than the animation duration.
+        if (id === this._lastAnimationId && elapsedTime < animationDuration) {
+          window.requestAnimationFrame(animation);
```
Comment (jelbourn) [inline src/components/progress-circle/progress-circle.ts]: You shouldn't need to use `window.` anywhere. Just `requestAnimationFrame` or `setInterval` etc.

File: src/components/progress-circle/progress-circle.ts [hunk 11]
```diff
@@ -41,47 +84,172 @@ export class MdProgressCircle {
...
+  private static clamp(v: number) {
     return Math.max(0, Math.min(100, v));
   }
-}
 
 
+  /**
+   * Easing function for linear animation.
+   */
+  private static linearEase(t: number, b: number, c: number, d: number) {
```
Comment (jelbourn) [inline src/components/progress-circle/progress-circle.ts]: Isn't an easing function supposed to just be a function of `t`? What are the other arguments?

File: src/components/progress-circle/progress-circle.ts [hunk 12]
```diff
@@ -41,47 +84,172 @@ export class MdProgressCircle {
...
+
+  /**
+   * Easing function to match material design indeterminate animation.
+   */
+  private static materialEase(t: number, b: number, c: number, d: number) {
+    // via http://www.timotheegroleau.com/Flash/experiments/easing_function_generator.htm
+    // with settings of [0, 0, 1, 1]
+    var ts = (t /= d) * t;
+    var tc = ts * t;
+    return b + c * (6 * tc * ts + -15 * ts * ts + 10 * tc);
```
Comment (jelbourn) [inline src/components/progress-circle/progress-circle.ts]: Even with the link, we need some explanation of what's going on here

File: src/components/progress-circle/progress-circle.ts [hunk 13]
```diff
@@ -41,47 +84,172 @@ export class MdProgressCircle {
...
+
+    return (radius + (pathRadius * Math.cos(angleInRadians))) +
+      ',' + (radius + (pathRadius * Math.sin(angleInRadians)));
+  }
+
+
+  /**
+   * Returns the current timestamp either based on the performance global or a date object.
+   */
+  private static _animationNow() {
```
Comment (jelbourn) [inline src/components/progress-circle/progress-circle.ts]: This can just be a module function and you can call it `now`.

File: src/components/progress-circle/progress-circle.ts [hunk 14]
```diff
@@ -41,47 +84,172 @@ export class MdProgressCircle {
...
+      this._interdeterminateInterval = null;
     }
   }
 
 
+  /**
+   * Determines the path value to define the arc.
+   */
+  private static getSvgArc(current: number, rotation: number) {
+    // The angle can't be exactly 360, because the arc becomes hidden.
```
Comment (jelbourn) [inline src/components/progress-circle/progress-circle.ts]: Also need high-level description of what's going on here.

File: src/components/progress-circle/progress-circle.ts (hunk 9 above)
Comment (jelbourn) [inline src/components/progress-circle/progress-circle.ts]: Rather than using the type `Function`, it should be a type with the function signature of the easing function.

File: src/components/progress-circle/progress-circle.ts [hunk 15]
```diff
@@ -1,13 +1,19 @@
...
+  ChangeDetectorRef,
+  ChangeDetectionStrategy,
   Input
 } from 'angular2/core';
-
+import {isPresent} from 'angular2/src/facade/lang';
 
 // TODO(josephperrott): Benchpress tests.
 
+let DEGREE_IN_RADIANS = Math.PI / 180;
```
Comment (robertmesserle) [inline src/components/progress-circle/progress-circle.ts]: Should this be a const?

Comment (robertmesserle) [inline src/components/progress-circle/progress-circle.ts]: Also, should all of these maybe be constants?  I only picked this one because it's all caps.

File: src/components/progress-circle/progress-circle.ts [hunk 16]
```diff
@@ -25,14 +31,51 @@ import {
...
+
+  /**
+   * The id of the indeterminate interval.
+   */
+  private _interdeterminateInterval: number;
+
+  /**
+   * A reference to the change detector.
+   */
+  private _ref: ChangeDetectorRef;
```
Comment (robertmesserle) [inline src/components/progress-circle/progress-circle.ts]: Maybe a more descriptive name?

File: src/components/progress-circle/progress-circle.ts [hunk 17]
```diff
@@ -41,47 +84,172 @@ export class MdProgressCircle {
...
    */
-  strokeDashOffset() {
-    // To determine how far the offset should be, we multiple the current percentage by the
-    // total circumference.
+  private _renderCircle(animateFrom: number, animateTo: number, ease: Function,
+                        duration: number, rotation: number) {
+    let id = ++this._lastAnimationId;
+    let startTime = MdProgressCircle._animationNow();
+    let changeInValue = animateTo - animateFrom;
+    let animationDuration = duration || 225;
```
Comment (robertmesserle) [inline src/components/progress-circle/progress-circle.ts]: Should this be set as a default value in the arguments?  ie. `duration: number = 225`

Comment (robertmesserle) [inline src/components/progress-circle/progress-circle.ts]: Or do we even really need a default here on a private method?

File: src/components/progress-circle/progress-circle.ts (hunk 11 above)
Comment (robertmesserle) [inline src/components/progress-circle/progress-circle.ts]: This looks like it was modeled after one of the Robert Penner easing functions:
`// t: current time, b: begInnIng value, c: change In value, d: duration`

Maybe expand these names to make it more obvious what is going on.

File: src/components/progress-circle/progress-circle.ts [hunk 18]
```diff
@@ -41,47 +84,172 @@ export class MdProgressCircle {
...
+    return (radius + (pathRadius * Math.cos(angleInRadians))) +
+      ',' + (radius + (pathRadius * Math.sin(angleInRadians)));
+  }
+
+
+  /**
+   * Returns the current timestamp either based on the performance global or a date object.
+   */
+  private static _animationNow() {
+    if (window.performance) {
```
Comment (robertmesserle) [inline src/components/progress-circle/progress-circle.ts]: Would it make sense to check with `typeof performance !== 'undefined'` rather than checking `window`?

File: src/components/progress-circle/progress-circle.ts (hunk 10 above)
Comment (josephperrott) [inline src/components/progress-circle/progress-circle.ts]: Done.

File: src/components/progress-circle/progress-circle.ts (hunk 8 above)
Comment (josephperrott) [inline src/components/progress-circle/progress-circle.ts]: Done

File: src/components/progress-circle/progress-circle.ts (hunk 7 above)
Comment (josephperrott) [inline src/components/progress-circle/progress-circle.ts]: Done, just to confirm, we do support iOS safari 8.4 as this is the only browser that may fall in scope of supported browsers that does that have performance.now

File: src/components/progress-circle/progress-circle.ts (hunk 16 above)
Comment (josephperrott) [inline src/components/progress-circle/progress-circle.ts]: Done

File: src/components/progress-circle/progress-circle.scss (hunk 1 above)
Comment (josephperrott) [inline src/components/progress-circle/progress-circle.scss]: Reverted back

File: src/components/progress-circle/progress-circle.ts (hunk 2 above)
Comment (josephperrott) [inline src/components/progress-circle/progress-circle.ts]: Done

File: src/components/progress-circle/progress-circle.ts (hunk 3 above)
Comment (josephperrott) [inline src/components/progress-circle/progress-circle.ts]: Done

File: src/components/progress-circle/progress-circle.ts (hunk 4 above)
Comment (josephperrott) [inline src/components/progress-circle/progress-circle.ts]: Done

File: src/components/progress-circle/progress-circle.ts (hunk 5 above)
Comment (josephperrott) [inline src/components/progress-circle/progress-circle.ts]: Done

File: src/components/progress-circle/progress-circle.ts (hunk 6 above)
Comment (josephperrott) [inline src/components/progress-circle/progress-circle.ts]: Done

File: src/components/progress-circle/progress-circle.ts (hunk 11 above)
Comment (josephperrott) [inline src/components/progress-circle/progress-circle.ts]: Done, I moved to more descriptive params.

File: src/components/progress-circle/progress-circle.ts (hunk 13 above)
Comment (josephperrott) [inline src/components/progress-circle/progress-circle.ts]: Done

File: src/components/progress-circle/progress-circle.ts (hunk 15 above)
Comment (josephperrott) [inline src/components/progress-circle/progress-circle.ts]: Updated to move the appropriate one to consts

File: src/components/progress-circle/progress-circle.ts (hunk 17 above)
Comment (josephperrott) [inline src/components/progress-circle/progress-circle.ts]: True, we can be sure that it will always get the duration provided since its private.

**Comment by jelbourn** [comment]
@joeyperson  Now needs rebasing on my mega change #384. Let me know if you need any help with that (next week, of course).

**Comment by josephperrott** [comment]
Rebased.

File: src/components/progress-circle/progress-circle.ts [hunk 19]
```diff
@@ -41,48 +84,100 @@ export class MdProgressCircle {
...
-   * Setting the offset to a percentage of the total circumference of the circle, fills this
-   * percentage of the overall circumference of the circle.
+   * Animates the rendered circle.
    */
-  strokeDashOffset() {
-    // To determine how far the offset should be, we multiple the current percentage by the
-    // total circumference.
+  private _renderCircle(animateFrom: number, animateTo: number,
+                        ease: (currentTime: number, startValue: number,
+                               changeInValue: number, duration: number) => number,
```
Comment (jelbourn) [inline src/components/progress-circle/progress-circle.ts]: You should be able to alias this type so that it's a little more readable here:
```ts
type EasingFn = (currentTime: number, startValue: number, changeInValue: number, duration: number) => number
```
(see the `Dir` directive for an example)

File: src/components/progress-circle/progress-circle.ts [hunk 20]
```diff
@@ -41,48 +84,100 @@ export class MdProgressCircle {
...
-   *
-   * The stroke dash offset specifies the distance between dashes in the circle's stroke.
-   * Setting the offset to a percentage of the total circumference of the circle, fills this
-   * percentage of the overall circumference of the circle.
+   * Animates the rendered circle.
    */
-  strokeDashOffset() {
-    // To determine how far the offset should be, we multiple the current percentage by the
-    // total circumference.
+  private _renderCircle(animateFrom: number, animateTo: number,
```
Comment (jelbourn) [inline src/components/progress-circle/progress-circle.ts]: Could you add some `@param` descriptions for this function? I think it would be helpful since it's not drop-dead obvious what the parameters are. Check out the grid-list for examples.

File: src/components/progress-circle/progress-circle.ts [hunk 21]
```diff
@@ -41,48 +84,100 @@ export class MdProgressCircle {
...
+    let id = ++this._lastAnimationId;
+    let startTime = now();
+    let changeInValue = animateTo - animateFrom;
 
-    // The total circumference is calculated based on the radius we use, 45.
-    // PI * 2 * 45
-    return 251.3274 * (100 - this._value) / 100;
-  }
+    // No need to animate it if the values are the same
+    if (animateTo === animateFrom) {
```
Comment (jelbourn) [inline src/components/progress-circle/progress-circle.ts]: I think this fn would be a little more eadable if it were formatted like
```ts
    // No need to animate it if the values are the same
    if (animateTo === animateFrom) {
      this.currentPath = getSvgArc(animateTo, rotation);
      return;
    }

    requestAnimationFrame(now => {
      // ...
    });
```

Comment (jelbourn) [inline src/components/progress-circle/progress-circle.ts]: Oh, nevermind on the second part, I missed the re-use of the `animation` inside.

File: src/components/progress-circle/progress-circle.ts [hunk 22]
```diff
@@ -96,11 +191,93 @@ export class MdProgressCircle {
...
+  return changeInValue * currentTime / duration + startValue;
+}
+
+
+/**
+ * Easing function to match material design indeterminate animation.
+ */
+function materialEase(currentTime: number, startValue: number,
+                      changeInValue: number, duration: number) {
+  var ts = (currentTime /= duration) * currentTime;
```
Comment (jelbourn) [inline src/components/progress-circle/progress-circle.ts]: It's weird to use `/=` in an intermediate calculation.

File: src/components/progress-circle/progress-circle.ts [hunk 23]
```diff
@@ -41,48 +84,100 @@ export class MdProgressCircle {
...
+          ease(elapsedTime, animateFrom, changeInValue, duration),
+          rotation
+        );
 
-  /** Gets the progress value, returning the clamped value. */
-  get value() {
-    return this._value;
+        // Prevent overlapping animations by checking if a new animation has been called for and
+        // if the animation has lasted long than the animation duration.
+        if (id === this._lastAnimationId && elapsedTime < duration && pageVisible()) {
```
Comment (jelbourn) [inline src/components/progress-circle/progress-circle.ts]: Since you're gating on `document.hidden`, do you need to listen for `visibilitychange` events to restart later?

File: src/components/progress-circle/progress-circle.ts [hunk 24]
```diff
@@ -96,11 +191,93 @@ export class MdProgressCircle {
...
+  var tc = ts * currentTime;
+  return startValue + changeInValue * (6 * tc * ts + -15 * ts * ts + 10 * tc);
+}
+
+
+/**
+ * Determines the path value to define the arc.
+ */
+function getSvgArc(current: number, rotation: number) {
+  // The angle can't be exactly 360, because the arc becomes hidden.
```
Comment (jelbourn) [inline src/components/progress-circle/progress-circle.ts]: This function would benefit from some more explanation of what's going on at a high level.

File: src/components/progress-circle/progress-circle.ts [hunk 25]
```diff
@@ -41,48 +84,100 @@ export class MdProgressCircle {
...
+        // Prevent rotation from reaching Number.MAX_SAFE_INTEGER.
+        rotationStartPoint = (rotationStartPoint + end) % 100;
+        let temp = start;
+        start = -end;
+        end = -temp;
+      }
+    };
+
+    if (!this._interdeterminateInterval) {
+      // Note that this interval isn't supposed to trigger a digest.
```
Comment (jelbourn) [inline src/components/progress-circle/progress-circle.ts]: We don't say the word "digest" any more 😉

When a `setInterval` function runs, it _will_ trigger a round of change detection unless you explicitly inject the `NgZone` and use that to `runOutsideAngular`.

File: src/components/progress-circle/progress-circle.ts [hunk 26]
```diff
@@ -24,15 +38,44 @@ import {
...
+  private _changeDetectorRef: ChangeDetectorRef;
+
+  /** The current path value, representing the progres circle. */
+  private _currentPath: string;
+  get currentPath() {
+    return this._currentPath;
+  }
+  set currentPath(path: string) {
+    this._currentPath = path;
+    this._changeDetectorRef.markForCheck();
```
Comment (jelbourn) [inline src/components/progress-circle/progress-circle.ts]: Add explanation of why you need to do this.

File: src/components/progress-circle/progress-circle.ts [hunk 27]
```diff
@@ -96,11 +191,93 @@ export class MdProgressCircle {
...
+    largeArcFlag = endAngle >= -180 ? 0 : 1;
+  } else {
+    largeArcFlag = endAngle <= 180 ? 0 : 1;
+  }
+
+  return `M${start}A${pathRadius},${pathRadius} 0 ${largeArcFlag},${arcSweep} ${end}`;
+}
+
+function pageVisible() {
+  return !document.hidden;
```
Comment (jelbourn) [inline src/components/progress-circle/progress-circle.ts]: I'd call this function `isPageVisible`, but then you might as well just use `document.hidden` directly since it's not much shorter. Up to you.

File: src/components/progress-circle/progress-circle.ts [hunk 28]
```diff
@@ -96,11 +191,93 @@ export class MdProgressCircle {
...
+  var ts = (currentTime /= duration) * currentTime;
+  var tc = ts * currentTime;
+  return startValue + changeInValue * (6 * tc * ts + -15 * ts * ts + 10 * tc);
+}
+
+
+/**
+ * Determines the path value to define the arc.
+ */
+function getSvgArc(current: number, rotation: number) {
```
Comment (jelbourn) [inline src/components/progress-circle/progress-circle.ts]: Need more explanation about what the function is returning.

File: src/components/progress-circle/progress-circle.ts (hunk 22 above)
Comment (josephperrott) [inline src/components/progress-circle/progress-circle.ts]: Done.

File: src/components/progress-circle/progress-circle.ts (hunk 23 above)
Comment (josephperrott) [inline src/components/progress-circle/progress-circle.ts]: Sorry, didn't realize I left this in.  browsers already prevent the rendering when not visible, so this is redundant.

File: src/components/progress-circle/progress-circle.ts (hunk 25 above)
Comment (josephperrott) [inline src/components/progress-circle/progress-circle.ts]: Done

File: src/components/progress-circle/progress-circle.ts (hunk 27 above)
Comment (josephperrott) [inline src/components/progress-circle/progress-circle.ts]: Removed.

File: src/components/progress-circle/progress-circle.ts (hunk 26 above)
Comment (josephperrott) [inline src/components/progress-circle/progress-circle.ts]: Done.

File: src/components/progress-circle/progress-circle.ts (hunk 19 above)
Comment (josephperrott) [inline src/components/progress-circle/progress-circle.ts]: Done

File: src/components/progress-circle/progress-circle.ts (hunk 20 above)
Comment (josephperrott) [inline src/components/progress-circle/progress-circle.ts]: Done

File: src/components/progress-circle/progress-circle.ts [hunk 29]
```diff
@@ -41,48 +89,102 @@ export class MdProgressCircle {
...
+    if (m == 'indeterminate') {
+      this._startIndeterminateAnimation();
+    } else {
+      this._cleanupIndeterminateAnimation();
+    }
+    this._mode = m;
+  }
+  private _mode: ProgressCircleMode = 'determinate';
+
+  constructor(changeDetectorRef: ChangeDetectorRef) {
```
Comment (jelbourn) [inline src/components/progress-circle/progress-circle.ts]: You can shorten this to 
```ts
constructor(private _changeDetectorRef: ChangeDetectorRef) { }
```

File: src/components/progress-circle/progress-circle.ts [hunk 30]
```diff
@@ -96,11 +198,101 @@ export class MdProgressCircle {
...
+/**
+ * Easing function to match material design indeterminate animation.
+ */
+function materialEase(currentTime: number, startValue: number,
+                      changeInValue: number, duration: number) {
+  let time = currentTime / duration;
+  let timeCubed = Math.pow(time, 3);
+  let timeQuad = Math.pow(time, 4);
+  let timeQuint = Math.pow(time, 5);
+  return startValue + changeInValue * ((6 * timeQuint) +
```
Comment (jelbourn) [inline src/components/progress-circle/progress-circle.ts]: This whole expression will actually fit on one line (our max is 100 cols)

**Comment by jelbourn** [comment]
LGTM aside from a couple minor style nits.

File: src/components/progress-circle/progress-circle.ts (hunk 30 above)
Comment (josephperrott) [inline src/components/progress-circle/progress-circle.ts]: Done

File: src/components/progress-circle/progress-circle.ts (hunk 29 above)
Comment (josephperrott) [inline src/components/progress-circle/progress-circle.ts]: Done.
